### DISCUSSION TITLE
Core features for PlotWidget and breaking out widgets.py

### DISCUSSION THREAD

**Discussion by mcdo0486** [pymeasure/pymeasure#491] [General]
Hi,

This discussion comes out of the PR #461 which implements multiple plots to the PlotWidget class. Some other features were discussed that could be added to PlotWidget, including multiple live curves on same plot or a secondary y axis. 

This discussion of features prompts a new question: What are the core features of PlotWidget and when should additional features use sub-class PlotWidget or create new widgets classes? I think the #461 PR is stuck in part b/c that question isn't being answered.

I think #461 should be merged and PlotWidget should be "frozen" as far as features. I think multiple graphs, which one Y axis and different X axis, is a general enough feature that new users will want. Instead new graphical features should be implemented as new widgets, which can subclass PlotWIdget if needed. 

However, to my next point, if new GUI features should be implemented as new widgets then I also think widgets.py should be broken out to a directory with a list of widget modules.

I know I'd like to make a PR for dockable windows, using the Dock Area module from pyqtgraph, and I'd like to do that as a new sub classed widget of PlotWidget.

**Comment by msmttchr** [comment]
> It is possible to display two plots with different y-axes in the managed window such that quantities' dependence on eachother can be easily seen.

If this the objective of the PR, I would say that, in my opinion, is a much common case to plot the multiple curves in the same plot. This will have the advantage of using more effectively the GUI space. On the other hand, I reckon there could be cases where two (PR is correctly generic in the number of plots, but possibly going beyond two can become ineffective due to limited y space) independent graphs can be more effective even though I have some difficulty to focus them. For example, if x axis is the same for both graph, the effectiveness is limited ( you just see two or more graph with y axis limited in space). If x and y are independent for both graph, again I think this is possible, but what is the use case ? Could you please offer some examples ?

**Comment by mcdo0486** [comment]
That is a point taken about multiple curves, and I'll let the PR author @casvharen speak to the objective of PR.

However I am still wondering what the core features of PlotWidget should be. Maybe #491 is rejected and submitted as a separate widget class, that is a possible future. 

But if the project will encourage graphical changes come as new widget classes, then I think widgets.py should be split out into separate module files, similar to what pyqtgraph does. I think pymeasure would be heading in that direction eventually if not now, the file is 1.3K lines long atm.

**Reply by msmttchr** [reply]
Fair point

**Comment by Casvharen** [comment]
Yeah the reason I made this was for experiments where the X-axis would be different for both plots. More than 2 plots does make them a little small, but if you make the screen fullscreen, then 3 plots is still managable, and if even more plots were needed, a scrollbar could be implemented relatively easily. Allowing the plots to share the same X-axis is a way to solve the space issue a little, but I can't imagine how to easily and clearly allow both possibilities, and sacrificing a little Y space for the extra flexibility of separate X-axis seemed like the right choice.

**Reply by msmttchr** [reply]
Thanks for the clarification. I think having the same x will also have issues when zooming or paning, since the two graphs are not in sync.
When x are different, do you see use cases for live plotting or to examine results after experiment is over or both ?

**Reply by Casvharen** [reply]
There would be for both I think. For example if you do a measurement where you alter a variable in a sweep fashion so you go back and forth E.g. a magnetic field, and in addition you want to measure something as a function of time, having two separate x-axis can be very convenient. For analysis purposes I suppose it could always be convenient to have another plot at your disposal
